var/spack/repos/builtin/packages/python/package.py

```diff
@@ -248,6 +248,30 @@ def post_install(self):
             os.symlink(os.path.join(prefix.bin, 'python3-config'),
                        os.path.join(prefix.bin, 'python-config'))
 
+        # Check for silent hashlib install problem on Intel compilers.
+        if spec.satisfies('%intel'):
+            # Interestingly, this is an un-catchable exception with an
+            # exit code of 0, so doing something like
+            #
+            #     import sys
+            #     try:
+            #         import hashlib
+            #     except:
+            #         sys.exit(1)
+            #
+            # will not actually exit with `1`, it will be `0`.  The failed
+            # `import hashlib` will print the traceback to stderr though,
+            # so we check that `stderr` is the empty string.  If it is not,
+            # the installation failed.
+            import_error = self.command('-c', 'import hashlib', error=str)
```

Comment:
won't this be executed in the stage directory and will try to import the local files?

Reply:
Hmm.  To be honest I don't really know.  I was originally going to find `prefix.bin` joined with `python` but then I saw this:

https://github.com/spack/spack/blob/3701dd0257eecc9461462d15f964db5f5965b633/var/spack/repos/builtin/packages/python/package.py#L218-L222

So I thought that was the right way / easier because `self.command` is super convenient.  Though I honestly don't understand how it even works, specifically why I can do `self.command(arg1, arg2)` rather than `self.command()(arg1, arg2)` -- `self.command()` to return the `Executable` and then a second `(arg1, arg2)` to actually do the `__call__`.

@adamjstewart you know this package inside and out, any advice here?

Reply:
> I honestly don't understand how it even works, specifically why I can do `self.command(arg1, arg2)` rather than `self.command()(arg1, arg2)` -- `self.command()` to return the `Executable` and then a second `(arg1, arg2)` to actually do the `__call__`.

That's because `self.command` is a property, not a function. `self.command` is sufficient to return the `Executable`.

As for the stage directory question, I'm wondering if we shouldn't wrap this logic in a dedicated post-installation test. See `gdal` for an example. That package also attempts an import, but in a separate directory so it avoids importing the local files and instead imports the installed files. The downside is that it will only be run if `--test=root` is supplied. But if you don't want to test your installations, you always run the risk of a broken installation.